torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -23,6 +23,57 @@ using torch::autograd::variable_list;
 static constexpr char* kNumBackwardPasses = "num_current_backward_passes";
 static constexpr char* kNumAutogradContexts = "num_autograd_contexts";
 
+// This hook does 3 things:
+//   1. Accumuate the gard to RPC context.
+//   2. Call pre hooks of the original AccumulateGrad to modify the input grad.
+//   3. Call post hooks of the original AccumulateGrad.
+struct DistAccumulateGradCaptureHook
+    : GraphTask::ExecInfo::Capture::GradCaptureHook {
+  DistAccumulateGradCaptureHook(
+      std::shared_ptr<AccumulateGrad> accumulateGrad,
+      ContextPtr autogradContext)
+      : accumulateGrad_(std::move(accumulateGrad)),
+        autogradContext_(std::move(autogradContext)) {}
+
+  torch::Tensor operator()(const torch::Tensor& grad) override {
+    // It is possible that the grad is not defined since a separate
+    // invocation of the autograd engine on the same node might actually
+    // compute this gradient.
+    if (grad.defined()) {
+      autogradContext_->accumulateGrad(
+          accumulateGrad_->variable, grad, 1 /* num_expected_refs */);
+    }
+    // It's intended that pre/post hooks are still called even if the grad is
+    // undenfined here.
+    if (!accumulateGrad_->pre_hooks().empty()) {
+      variable_list inputGrads = {grad};
+      for (const auto& hook : accumulateGrad_->pre_hooks()) {
+        inputGrads = (*hook)(inputGrads);
+        // A pre-hook should return a varaible list of the same size as the
+        // function's inputs.
+        TORCH_INTERNAL_ASSERT(inputGrads.size() == 1);
+      }
+      grad = inputGrads[0];
```

Comment:
The input grad is modified by the pre-hooks. The intention is to keep the same semantic of pre-hooks. Let me if it shouldn't modify the input grad. @pritamdamania87  @albanD

Reply:
Yes the hooks modify the value that go to the next hook and then next steps.